### PR TITLE
AKS job cleanup

### PR DESCRIPTION
**Pull request by ilausuch**
https://progress.opensuse.org/issues/123502

### CODE REVIEW

**Comment by ilausuch** [comment]
Remains the test for aks

File: ocw/lib/aks.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+            AKS.__container_mgmt_client = None
+            AKS.__sp_credentials = None
+            AKS.__kubectl_client = {}
+        return AKS.__instances[vault_namespace]
+
+    def subscription(self) -> str:
+        return self.get_data('subscription_id')
+
+    def resource_group_name(self) -> str:
+        return self.get_data('resource_group_name')
```
Comment (asmorodskyi) [inline ocw/lib/aks.py:25]: AFAIK this files does not have `'resource_group_name'` at least the one which we have in prod certainly does not have it . Also this is single file which is global for single account and this field should be specific for certain single cluster

Comment (asmorodskyi) [reply ocw/lib/aks.py:25]: I think it won't work we need to find another way to get resource group name for AKS cluster

File: ocw/lib/aks.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+            resource_group = self.resource_group_name()
+
+            res = self.cmd_exec(f"az login --service-principal -u {self.get_data('client_id')} "
+                                f"-p {self.get_data('client_secret')} --tenant {self.get_data('tenant_id')}")
+            if res.returncode != 0:
+                raise RuntimeError(f"Cannot login to azure : {res.stderr}")
+
+            res = self.cmd_exec(f"az aks get-credentials --resource-group {resource_group} "
+                                f"--name {cluster_name} --file {kubeconfig}",
+                                aditional_env={"PUBLIC_CLOUD_K8S_CLUSTER": cluster_name,
```
[the remark this answers is not in the document]
Comment (ilausuch) [reply ocw/lib/aks.py:56]: I don't remember exactly, but what I remember is that without that this method doesn't work. Maybe can be set on arguments. I'll take a look

**Comment by ilausuch** [comment]
Now we have the test test_aks

**Comment by ilausuch** [comment]
Now the clusters are configured on section [k8sclusters]
e.g.
```
[k8sclusters]
providers = azure
azure-clusters = resource_group_name:cluster_name
```

File: ocw/lib/aks.py (hunk 1 above)
Comment (ilausuch) [reply ocw/lib/aks.py]: done

File: webui/PCWConfig.py [hunk 3]
```diff
@@ -87,6 +87,18 @@ def get_providers_for(feature: str, namespace: str):
         return ConfigFile().getList('{}.namespace.{}/providers'.format(feature, namespace),
                                     ConfigFile().getList('{}/providers'.format(feature), ['EC2', 'AZURE', 'GCE']))
 
+    @staticmethod
+    def get_k8s_clusters_for_provider(provider: str) -> list:
```
Comment (asmorodskyi) [inline webui/PCWConfig.py:91]: test_pcwconfig is missing this ;)

File: webui/PCWConfig.py [hunk 4]
```diff
@@ -87,6 +87,18 @@ def get_providers_for(feature: str, namespace: str):
...
+        result = []
+        providers = ConfigFile().get("k8sclusters/providers")
+        if provider in providers:
+            clusters = ConfigFile().get(f"k8sclusters/{provider}-clusters").strip()
```
Comment (asmorodskyi) [inline webui/PCWConfig.py:95]: I don't think `strip` is necessary here : 
1. please read carefully `ConfigFile().get` we getting this string from https://docs.python.org/3/library/configparser.html  I hardly believe that they not handling leading/trailing whitespaces in their code 
2. if they really don't ( but I need some proves to believe in that :) ) than we should handle this **within**  ConfigFile().get . Certainly not in every place where we calling `ConfigFile().get`

File: webui/PCWConfig.py [hunk 5]
```diff
@@ -87,6 +87,18 @@ def get_providers_for(feature: str, namespace: str):
...
+            for cluster in clusters.split(','):
+                cluster = cluster.strip()
+                resource_group, cluster_name = cluster.split(':')
```
Comment (asmorodskyi) [inline webui/PCWConfig.py:98]: here we need some verification that string was formatted properly and if not we must throw exception here

File: ocw/lib/aks.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (asmorodskyi) [reply ocw/lib/aks.py:60]: not that I am against having this variable here , if it is really needed I am fine with that but let's double check that and leave some comment explaining why we need it here so future us will have easier life on figuring out what is going on here

[the remark this answers is not in the document]
Comment (pdostal) [reply ocw/lib/aks.py:60]: And maybe commend it somehow.

File: webui/PCWConfig.py (hunk 4 above)
Comment (ilausuch) [reply webui/PCWConfig.py:95]: Checked that is not necessary. I tested this and works
`azure-clusters = area1:cluster1, area1:cluster2 `

Comment (ilausuch) [reply webui/PCWConfig.py:95]: however the next strip is really necessary `cluster = cluster.strip()`

File: ocw/lib/aks.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ilausuch) [reply ocw/lib/aks.py:60]: I rechecked this and seems not necessary

File: ocw/lib/aks.py [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+from ocw.lib.provider import Provider
+from ocw.lib.k8s import clean_jobs
+from webui.PCWConfig import PCWConfig
+
+
+class AKS(Provider):
+    __instances = {}
+    default_region: str = 'eu-central-1'
+
+    def __new__(cls, vault_namespace):
```
[the remark this answers is not in the document]
Comment (ilausuch) [reply ocw/lib/aks.py:15]: done

File: webui/PCWConfig.py (hunk 4 above)
Comment (ilausuch) [reply webui/PCWConfig.py]: done

File: webui/PCWConfig.py (hunk 5 above)
Comment (ilausuch) [reply webui/PCWConfig.py:100]: done

**Comment by ilausuch** [comment]
WIP for the testing test_pcwconfig

**Comment by asmorodskyi** [comment]
please rebase on latest master so you will get also codecov report in your PR

File: ocw/lib/aks.py [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+        return self.__container_mgmt_client
+
+    def resource_mgmt_client(self):
+        if self.__resource_mgmt_client is None:
+            self.__resoure_mgmt_client = ResourceManagementClient(
+                self.sp_credentials(), self.subscription())
+        return self.__resoure_mgmt_client
+
+    def get_clusters(self):
+        return self.container_mgmt_client().managed_clusters.list_by_resource_group(self.resource_group_name())
```
Comment (asmorodskyi) [inline ocw/lib/aks.py:46]: I failed to find implementation of `resource_group_name` method implementation

File: tests/test_aks.py [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+    assert aks_patch.kubectl_client("group", "cluster") == mocked_client1
+
+    # Check that the client is reused
+    mocked_client2 = MockedKubernetesClient(1)
+    monkeypatch.setattr(kubernetes, 'client', mocked_client2)
+    assert aks_patch.kubectl_client("group", "cluster") == mocked_client1
+
+
+def test_cleanup_k8s_jobs(aks_patch, monkeypatch):
+    monkeypatch.setattr(kubernetes, 'config', MockedKubernetesConfig())
```
Comment (asmorodskyi) [inline tests/test_aks.py:40]: this line presented in both tests hence you can move it into `aks_patch`

File: ocw/lib/aks.py (hunk 7 above)
Comment (ilausuch) [reply ocw/lib/aks.py]: Thanks, done

File: tests/test_aks.py (hunk 8 above)
Comment (ilausuch) [reply tests/test_aks.py]: done

**Comment by asmorodskyi** [comment]
generally all looks good , I am confident with code , so let's add more tests for PCWConfig and we are done

File: tests/test_pcwconfig.py [hunk 9]
```diff
@@ -140,3 +140,15 @@ def test_getBoolean_namespace_but_not_defined(pcw_file):
     providers = azure
     """)
     assert PCWConfig.getBoolean('feature/bool_property', 'random_namespace')
+
```
[the remark this answers is not in the document]
Comment (ilausuch) [reply tests/test_pcwconfig.py:143]: done

File: webui/PCWConfig.py [hunk 10]
```diff
@@ -87,6 +87,26 @@ def get_providers_for(feature: str, namespace: str):
...
+    def get_k8s_providers() -> list:
+        providers = ConfigFile().get("k8sclusters/providers")
+        return [provider.strip() for provider in providers.split(',')]
+
+    @staticmethod
+    def get_k8s_clusters_for_provider(provider: str) -> list:
+        result = []
+        providers = ConfigFile().get("k8sclusters/providers")
+        if provider in providers:
+            clusters = ConfigFile().get(f"k8sclusters/{provider}-clusters")
```
Comment (asmorodskyi) [inline webui/PCWConfig.py:100]:
```suggestion
            clusters = ConfigFile().get(f"k8sclusters.namespace.{namespace}/{provider}-clusters")
```

File: ocw/lib/aks.py [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+                                   f"for resource-group {resource_group} : {res.stderr}")
+
+            kubernetes.config.load_kube_config(config_file=kubeconfig)
+            self.__kubectl_client[cluster_name] = kubernetes.client.BatchV1Api()
+
+        return self.__kubectl_client[cluster_name]
+
+    def cleanup_k8s_jobs(self):
+        clusters = PCWConfig.get_k8s_clusters_for_provider(self._namespace, "azure")
+        print(clusters)
```
Comment (asmorodskyi) [inline ocw/lib/aks.py:68]: I assume it is debugging leftover

File: webui/PCWConfig.py [hunk 12]
```diff
@@ -87,6 +87,24 @@ def get_providers_for(feature: str, namespace: str):
         return ConfigFile().getList('{}.namespace.{}/providers'.format(feature, namespace),
                                     ConfigFile().getList('{}/providers'.format(feature), ['EC2', 'AZURE', 'GCE']))
 
+    @staticmethod
+    def get_k8s_providers() -> list:
```
Comment (asmorodskyi) [inline webui/PCWConfig.py:91]: we can delete it now

Comment (ilausuch) [reply webui/PCWConfig.py]: Yes, I did yesterday but I didnt' commit the changes :)

File: ocw/lib/aks.py (hunk 11 above)
Comment (ilausuch) [reply ocw/lib/aks.py:68]: yes!

**Review by pdostal — APPROVED** [review]
Looks good to me :+1:

**Pull request merged**